vms/proposervm/proposer/windower.go

```diff
@@ -148,3 +167,116 @@ func (w *windower) Delay(ctx context.Context, chainHeight, pChainHeight uint64,
+	blockHeight,
+	pChainHeight uint64,
+	parentBlockTime,
+	blockTime time.Time,
+) (ids.NodeID, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	var (
+		slot             = uint64(blockTime.Sub(parentBlockTime) / WindowDuration)
+		seed             = blockHeight ^ bits.Reverse64(slot) ^ w.chainSource
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+	)
+
+	sampler := sampler.NewDeterministicWeightedWithoutReplacement(source)
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	source.Seed(seed)
+	indices, err := sampler.Sample(1)
+	if err != nil {
+		return ids.EmptyNodeID, fmt.Errorf("%w, %w", err, ErrNoProposersAvailable)
+	}
+	return validators[indices[0]].id, nil
+}
+
+func (w *windower) MinDelayForProposer(
+	ctx context.Context,
+	chainHeight,
+	pChainHeight uint64,
+	parentBlockTime time.Time,
+	nodeID ids.NodeID,
+	initialSlot uint64,
+) (time.Duration, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return 0, err
+	}
+
+	var (
+		maxSlot = initialSlot + MaxLookAheadSlots
+
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+		sampler          = sampler.NewDeterministicWeightedWithoutReplacement(source)
+	)
+
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return 0, err
+	}
+
+	for slot := initialSlot; slot < maxSlot; slot++ {
+		seed := chainHeight ^ bits.Reverse64(slot) ^ w.chainSource
+		source.Seed(seed)
+		indices, err := sampler.Sample(1)
+		if err != nil {
+			return 0, fmt.Errorf("%w, %w", err, ErrNoProposersAvailable)
+		}
+
+		if validators[indices[0]].id == nodeID {
+			return time.Duration(slot) * WindowDuration, nil
+		}
+	}
+
+	// no slots scheduled for the max window we inspect. Return max delay
+	return time.Duration(maxSlot) * WindowDuration, nil
+}
+
+func (w *windower) sortedValidators(ctx context.Context, pChainHeight uint64) ([]validatorData, error) {
+	// get the validator set by the p-chain height
+	validatorsMap, err := w.state.GetValidatorSet(ctx, pChainHeight, w.subnetID)
+	if err != nil {
+		return nil, err
+	}
+
+	// convert the map of validators to a slice
+	validators := make([]validatorData, 0, len(validatorsMap))
+	for k, v := range validatorsMap {
+		validators = append(validators, validatorData{
+			id:     k,
+			weight: v.Weight,
+		})
+	}
+
+	// canonically sort validators
+	// Note: validators are sorted by ID, sorting by weight would not create a
+	// canonically sorted list
+	utils.Sort(validators)
+
+	return validators, nil
+}
+
+func validatorsToWeight(validators []validatorData) []uint64 {
+	weights := make([]uint64, len(validators))
+	for i, validator := range validators {
+		weights[i] = validator.weight
+	}
+	return weights
+}
+
+func TimeToSlot(baseTime, targetTime time.Time) uint64 {
+	if !targetTime.After(baseTime) {
```

Comment:
Unless we really care about how time equality is handled - I think we should avoid negating the condition.
```suggestion
	if targetTime.Before(baseTime) {
```